File: scripts/measure.ts
/**
 * Measures compression over the recorded prompts. `--savings` prints what it
 * saves, per level and per request; the corpus is the one the tests use, so
 * these are the numbers the README quotes. `--performance` prints how long it
 * takes. Savings is the default.
 */
import type { Level } from '../src/compression/levels.js';
import { LEVEL_NAMES } from '../src/compression/levels.js';
import { ALL_SCOPES } from '../src/ir/walk.js';
import { toIR } from '../src/adapters/anthropic/to-ir.js';
import { runPipeline } from '../src/compression/pipeline.js';
import { accountFor } from '../src/telemetry/accounting.js';
import { PROMPT_FIXTURES } from '../test/fixtures/prompts.js';

type Measurement = {
  name: string;
  tokensBefore: number;
  tokensAfter: number;
  ratio: number;
  proseShare: number;
};

function measure(body: Record<string, unknown>, level: Level): Measurement {
  const result = runPipeline({
    request: toIR(body),
    level,
    scopes: ALL_SCOPES,
    cacheMode: 'ignore',
  });
  const accounting = accountFor(result.stats);
  return {
    name: '',
    tokensBefore: accounting.tokensBefore,
    tokensAfter: accounting.tokensAfter,
    ratio: accounting.ratio,
    proseShare: result.stats.charsProse / result.stats.charsBefore,
  };
}

function percent(value: number): string {
  return `${(value * 100).toFixed(1)}%`;
}

function count(value: number): string {
  return value.toLocaleString('en-US');
}

function printCorpusTotals(): void {
  console.log('corpus');
  for (const level of LEVEL_NAMES) {
    let before = 0;
    let after = 0;
    for (const fixture of PROMPT_FIXTURES) {
      const measurement = measure(fixture.body, level);
      before += measurement.tokensBefore;
      after += measurement.tokensAfter;
    }
    const saved = percent((after - before) / before);
    console.log(`  ${level.padEnd(9)} ${count(before)} → ${count(after)} tok  ${saved}`);
  }
}

function printPerRequest(level: Level): void {
  console.log(`\nby request, at ${level} levels`);
  const rows = PROMPT_FIXTURES.map((fixture) => ({
    ...measure(fixture.body, level),
    name: fixture.name,
  })).sort((left, right) => right.proseShare - left.proseShare);

  for (const row of rows) {
    const prose = `${(row.proseShare * 100).toFixed(0)}%`.padStart(4);
    const saved = percent(-row.ratio).padStart(7);
    console.log(`  ${row.name.slice(0, 48).padEnd(48)} ${prose} prose ${saved}`);
  }
}

/**
 * Trials per timed subject. `compromise` tags on first use and the JIT needs a
 * few passes to settle, so a single run measures warmup rather than the work.
 */
const WARMUP_RUNS = 5;
const TIMED_RUNS = 20;

type Timing = {
  name: string;
  median: number;
  min: number;
  max: number;
  /** Prose characters the run had to classify, the driver of the cost. */
  charsProse: number;
};

/**
 * Times the pipeline alone. `toIR` runs once up front and its result is reused,
 * because parsing the wire format is the server's cost on every request whether
 * or not compression is on, and including it would understate the share
 * compression adds.
 */
function time(name: string, body: Record<string, unknown>, level: Level): Timing {
  const request = toIR(body);
  const run = (): number => {
    const started = performance.now();
    runPipeline({ request, level, scopes: ALL_SCOPES, cacheMode: 'ignore' });
    return performance.now() - started;
  };

  for (let index = 0; index < WARMUP_RUNS; index += 1) run();

  const samples: number[] = [];
  for (let index = 0; index < TIMED_RUNS; index += 1) samples.push(run());
  samples.sort((left, right) => left - right);

  const stats = runPipeline({ request, level, scopes: ALL_SCOPES, cacheMode: 'ignore' }).stats;
  return {
    name,
    // Median, not mean: a GC pause during one trial should not move the number.
    median: samples[Math.floor(samples.length / 2)] ?? 0,
    min: samples[0] ?? 0,
    max: samples[samples.length - 1] ?? 0,
    charsProse: stats.charsProse,
  };
}

function milliseconds(value: number): string {
  return `${value.toFixed(2)}ms`;
}

function printLevelTimings(): void {
  console.log(`corpus, ${TIMED_RUNS} runs each after ${WARMUP_RUNS} warmup`);
  for (const level of LEVEL_NAMES) {
    let median = 0;
    let charsProse = 0;
    for (const fixture of PROMPT_FIXTURES) {
      const timing = time(fixture.name, fixture.body, level);
      median += timing.median;
      charsProse += timing.charsProse;
    }
    // Throughput over prose characters, which is what the classifier walks.
    const perSecond = charsProse / (median / 1000);
    console.log(
      `  ${level.padEnd(9)} ${milliseconds(median).padStart(9)}` +
        `  ${count(Math.round(perSecond / 1000))}k prose chars/s`,
    );
  }
}

function printPerRequestTimings(level: Level): void {
  console.log(`\nby request, at ${level} levels`);
  const rows = PROMPT_FIXTURES.map((fixture) =>
    time(fixture.name, fixture.body, level),
  ).sort((left, right) => right.median - left.median);

  for (const row of rows) {
    const median = milliseconds(row.median).padStart(9);
    const spread = `${milliseconds(row.min)}–${milliseconds(row.max)}`.padStart(18);
    console.log(`  ${row.name.slice(0, 48).padEnd(48)} ${median} ${spread}`);
  }
}

const MODES = ['savings', 'performance'] as const;
type Mode = (typeof MODES)[number];

function parseMode(argv: readonly string[]): Mode {
  const flags = argv.filter((argument) => argument.startsWith('--'));
  if (flags.length === 0) return 'savings';
  if (flags.length > 1) {
    console.error(`measure: give one of ${MODES.map((mode) => `--${mode}`).join(', ')}`);
    process.exit(1);
  }
  const requested = flags[0]?.slice(2);
  const mode = MODES.find((candidate) => candidate === requested);
  if (mode === undefined) {
    console.error(
      `measure: unknown flag "${flags[0]}", expected ${MODES.map((name) => `--${name}`).join(' or ')}`,
    );
    process.exit(1);
  }
  return mode;
}

if (parseMode(process.argv.slice(2)) === 'performance') {
  printLevelTimings();
  printPerRequestTimings('caveman');
} else {
  printCorpusTotals();
  printPerRequest('caveman');
}

File: internal/cli/measure.go
package cli

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/carlelieser/caveman/internal/adapters/anthropic"
	"github.com/carlelieser/caveman/internal/compress"
	"github.com/carlelieser/caveman/internal/ir"
	"github.com/carlelieser/caveman/internal/telemetry"
)

// promptFixtures are the recorded bodies carrying prompts at the length real
// traffic arrives in, ordered as the corpus was recorded. The structural
// fixtures in the same directory cover wire shapes and carry too little text to
// measure, so they are named out rather than globbed in.
var promptFixtures = []string{
	"beginner rambling about a react state bug",
	"formal bug report with stack trace and package manifest",
	"terse expert asking about a postgres plan",
	"long support-agent system prompt with a casual question",
	"six-turn debugging conversation that goes sideways",
	"tool conversation with a large json tool_result",
	"tool conversation with a wall of log output",
	"code review request that is mostly a pasted diff",
	"casual one-liner with a pasted link",
	"dense prose with no code at all",
	"migration planning with a cached repo-context prefix",
	"long meandering devops question with a compose file",
	"four-turn api design discussion with mixed prose and snippets",
}

type corpusEntry struct {
	name    string
	request ir.Request
}

// corpusDir is where the recorded bodies live relative to the repository root.
// Measuring reads them from disk rather than embedding them, so the corpus the
// tests gate against and the corpus the numbers come from cannot drift apart.
const corpusDir = "testdata/golden/fixtures"

func loadCorpus(root string) ([]corpusEntry, error) {
	dir := filepath.Join(root, corpusDir)
	indexBytes, err := os.ReadFile(filepath.Join(dir, "index.json"))
	if err != nil {
		return nil, fmt.Errorf("reading the corpus index: %w", err)
	}
	var index []struct {
		Name string `json:"name"`
		File string `json:"file"`
	}
	if err := json.Unmarshal(indexBytes, &index); err != nil {
		return nil, fmt.Errorf("parsing the corpus index: %w", err)
	}
	files := map[string]string{}
	for _, entry := range index {
		files[entry.Name] = entry.File
	}

	entries := make([]corpusEntry, 0, len(promptFixtures))
	for _, name := range promptFixtures {
		file, ok := files[name]
		if !ok {
			return nil, fmt.Errorf("the corpus index does not name %q", name)
		}
		raw, err := os.ReadFile(filepath.Join(dir, file))
		if err != nil {
			return nil, fmt.Errorf("reading %s: %w", file, err)
		}
		value, err := ir.Unmarshal(raw)
		if err != nil {
			return nil, fmt.Errorf("parsing %s: %w", file, err)
		}
		body, ok := value.(*ir.Object)
		if !ok {
			return nil, fmt.Errorf("%s: the body is not an object", file)
		}
		entries = append(entries, corpusEntry{name: name, request: anthropic.ToIR(body)})
	}
	return entries, nil
}

type measurement struct {
	name         string
	tokensBefore int
	tokensAfter  int
	ratio        float64
	proseShare   float64
}

func measureOne(entry corpusEntry, level compress.Level) measurement {
	result := compress.RunPipeline(compress.PipelineRequest{
		Request:   entry.request,
		Level:     level,
		Scopes:    ir.AllScopes,
		CacheMode: compress.CacheIgnore,
	})
	accounting := telemetry.AccountFor(result.Stats)
	share := 0.0
	if result.Stats.CharsBefore > 0 {
		share = float64(result.Stats.CharsProse) / float64(result.Stats.CharsBefore)
	}
	return measurement{
		name:         entry.name,
		tokensBefore: accounting.TokensBefore,
		tokensAfter:  accounting.TokensAfter,
		ratio:        accounting.Ratio,
		proseShare:   share,
	}
}

func percent(value float64) string {
	return fmt.Sprintf("%.1f%%", value*100)
}

// group writes a count with en-US thousands separators, which is how the
// savings log prints its own totals.
func group(value int) string {
	digits := fmt.Sprintf("%d", value)
	if len(digits) <= 3 {
		return digits
	}
	out := strings.Builder{}
	lead := len(digits) % 3
	if lead > 0 {
		out.WriteString(digits[:lead])
	}
	for at := lead; at < len(digits); at += 3 {
		if out.Len() > 0 {
			out.WriteByte(',')
		}
		out.WriteString(digits[at : at+3])
	}
	return out.String()
}

func padRight(text string, width int) string {
	if len(text) >= width {
		return text
	}
	return text + strings.Repeat(" ", width-len(text))
}

func padLeft(text string, width int) string {
	if len(text) >= width {
		return text
	}
	return strings.Repeat(" ", width-len(text)) + text
}

func truncate(text string, width int) string {
	if len(text) <= width {
		return text
	}
	return text[:width]
}

func (c *CLI) printCorpusTotals(corpus []corpusEntry) {
	c.streams.say("corpus")
	for _, level := range compress.LevelNames {
		before, after := 0, 0
		for _, entry := range corpus {
			m := measureOne(entry, level)
			before += m.tokensBefore
			after += m.tokensAfter
		}
		saved := 0.0
		if before > 0 {
			saved = float64(after-before) / float64(before)
		}
		c.streams.say("  %s %s → %s tok  %s",
			padRight(string(level), 9), group(before), group(after), percent(saved))
	}
}

func (c *CLI) printPerRequest(corpus []corpusEntry, level compress.Level) {
	c.streams.say("")
	c.streams.say("by request, at %s levels", level)
	rows := make([]measurement, 0, len(corpus))
	for _, entry := range corpus {
		rows = append(rows, measureOne(entry, level))
	}
	sort.SliceStable(rows, func(left, right int) bool {
		return rows[left].proseShare > rows[right].proseShare
	})
	for _, row := range rows {
		prose := padLeft(fmt.Sprintf("%.0f%%", row.proseShare*100), 4)
		saved := padLeft(percent(-row.ratio), 7)
		c.streams.say("  %s %s prose %s", padRight(truncate(row.name, 48), 48), prose, saved)
	}
}

// Trials per timed subject. A single run measures a cold cache rather than the
// work, so the first few are thrown away and the reported figure is a median.
const (
	warmupRuns = 5
	timedRuns  = 20
)

type timing struct {
	name       string
	median     time.Duration
	low        time.Duration
	high       time.Duration
	charsProse int
}

// timeOne times the pipeline alone. Parsing the wire format is the server's
// cost on every request whether or not compression is on, so ToIR runs once
// during loading and its result is reused; including it would understate the
// share compression adds.
func timeOne(entry corpusEntry, level compress.Level) timing {
	request := compress.PipelineRequest{
		Request:   entry.request,
		Level:     level,
		Scopes:    ir.AllScopes,
		CacheMode: compress.CacheIgnore,
	}
	run := func() (time.Duration, int) {
		started := time.Now()
		result := compress.RunPipeline(request)
		return time.Since(started), result.Stats.CharsProse
	}
	for index := 0; index < warmupRuns; index++ {
		run()
	}
	samples := make([]time.Duration, 0, timedRuns)
	charsProse := 0
	for index := 0; index < timedRuns; index++ {
		elapsed, prose := run()
		samples = append(samples, elapsed)
		charsProse = prose
	}
	sort.Slice(samples, func(left, right int) bool { return samples[left] < samples[right] })
	return timing{
		name:       entry.name,
		median:     samples[len(samples)/2],
		low:        samples[0],
		high:       samples[len(samples)-1],
		charsProse: charsProse,
	}
}

func milliseconds(value time.Duration) string {
	return fmt.Sprintf("%.2fms", float64(value.Nanoseconds())/1e6)
}

func (c *CLI) printLevelTimings(corpus []corpusEntry) {
	c.streams.say("corpus, %d runs each after %d warmup", timedRuns, warmupRuns)
	for _, level := range compress.LevelNames {
		total := time.Duration(0)
		charsProse := 0
		for _, entry := range corpus {
			t := timeOne(entry, level)
			total += t.median
			charsProse += t.charsProse
		}
		perSecond := 0.0
		if total > 0 {
			perSecond = float64(charsProse) / total.Seconds()
		}
		c.streams.say("  %s %s  %sk prose chars/s",
			padRight(string(level), 9), padLeft(milliseconds(total), 9),
			group(int(perSecond/1000+0.5)))
	}
}

func (c *CLI) printPerRequestTimings(corpus []corpusEntry, level compress.Level) {
	c.streams.say("")
	c.streams.say("by request, at %s levels", level)
	rows := make([]timing, 0, len(corpus))
	for _, entry := range corpus {
		rows = append(rows, timeOne(entry, level))
	}
	sort.SliceStable(rows, func(left, right int) bool { return rows[left].median > rows[right].median })
	for _, row := range rows {
		spread := milliseconds(row.low) + "–" + milliseconds(row.high)
		c.streams.say("  %s %s %s", padRight(truncate(row.name, 48), 48),
			padLeft(milliseconds(row.median), 9), padLeft(spread, 18))
	}
}

// measure reports what compression saves over the recorded corpus, or how long
// it takes. Savings is the default.
func (c *CLI) measure(argv []string) *exitError {
	mode := "savings"
	for _, argument := range argv {
		switch argument {
		case "--savings", "--performance":
			mode = strings.TrimPrefix(argument, "--")
		default:
			return die(ExitUsage, "caveman: measure takes --savings or --performance, not %q", argument)
		}
	}

	root, err := os.Getwd()
	if err != nil {
		return die(ExitFailure, "caveman: %s", err)
	}
	corpus, err := loadCorpus(root)
	if err != nil {
		return die(ExitFailure, "caveman: %s", err)
	}

	if mode == "performance" {
		c.printLevelTimings(corpus)
		c.printPerRequestTimings(corpus, compress.LevelCaveman)
		return nil
	}
	c.printCorpusTotals(corpus)
	c.printPerRequest(corpus, compress.LevelCaveman)
	return nil
}
